python/tank_vendor/shotgun_authentication/session_cache.py

```diff
@@ -0,0 +1,378 @@
+import logging
+
+logger = logging.getLogger("shotgun_authentication").getChild("session_cache")
+
+def _get_cache_location():
+    """
+    Returns an OS specific cache location.
+
+    :returns: Path to the OS specific cache folder.
+    """
+    if sys.platform == "darwin":
+        root = os.path.expanduser("~/Library/Caches/Shotgun")
+    elif sys.platform == "win32":
+        root = os.path.join(os.environ["APPDATA"], "Shotgun")
+    elif sys.platform.startswith("linux"):
+        root = os.path.expanduser("~/.shotgun")
+    return root
+
+
+def _get_site_cache_location(base_url):
+    """
+    Returns the location of the site cache root based on a site.
+
+    :param base_url: Site we need to compute the root path for.
+
+    :returns: An absolute path to the site cache root.
+    """
+    return os.path.join(
+        _get_cache_location(),
+        # get site only; https://www.foo.com:8080 -> www.foo.com
+        urlparse.urlparse(base_url)[1].split(":")[0]
+    )
+
+
+def _get_global_authentication_file_location():
+    """
+    Returns the location of the authentication file on disk. This file
+    stores authentication related information for all sites. At this moment,
+    the file stores only the current host.
+
+    :returns: Path to the login information.
+    """
+    return os.path.join(
+        _get_cache_location(),
+        "authentication.yml"
+    )
+
+
+def _get_site_authentication_file_location(base_url):
+    """
+    Returns the location of the users file on disk for a specific site.
+
+    :param base_url: The site we want the login information for.
+
+    :returns: Path to the login information.
+    """
+    return os.path.join(
+        _get_site_cache_location(base_url),
+        "authentication.yml"
+    )
+
+
+def _ensure_folder_for_file(filepath):
+    """
+    Makes sure the folder exists for a given file.
+
+    :param filepath: Path to the file we want to make sure the parent directory
+                     exists.
+
+    :returns: The path to the file.
+    """
+
+    folder, _ = os.path.split(filepath)
+    if not os.path.exists(folder):
+        old_umask = os.umask(0077)
+        try:
+            os.makedirs(folder, 0700)
+        finally:
+            os.umask(old_umask)
+    return filepath
+
+
+def _try_load_yaml_file(file_path):
+    """
+    Loads a yaml file.
+
+    :param file_path: The yaml file to load.
+
+    :returns: The dictionary for this yaml file. If the file doesn't exist or is
+              corrupted, returns an empty dictionary.
+    """
+
+    if not os.path.exists(file_path):
+        logger.debug("Yaml file missing: %s" % file_path)
+        return {}
+    try:
+        # Open the file and read it.
+        with open(file_path, "r") as config_file:
+            return yaml.load(config_file)
+    except yaml.YAMLError:
+        # Return to the beginning
+        config_file.seek(0)
+        logger.error("File '%s' is corrupted!" % file_path)
```

Comment:
How about also pulling in the error message from the exception here? e.g. `logger.error("File '%s' is corrupted: %s" % (file_path, err))`

And what about other errors outside `yaml.YAMLError`? What if the file is held in some sort of lock for example or there is some other I/O error? Do we want it to return {} or to raise?